File: backend/app/schemas/ichimoku.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.market import HistoricalBar


class IchimokuPoint(BaseModel):
    timestamp: datetime
    tenkan: float | None = None
    kijun: float | None = None
    senkou_a: float | None = None
    senkou_b: float | None = None
    chikou: float | None = None
    close: float | None = None


class WavePivot(BaseModel):
    timestamp: datetime
    price: float


class WaveTargetSet(BaseModel):
    pivot_a: WavePivot
    pivot_b: WavePivot
    pivot_c: WavePivot
    v_target: float
    n_target: float
    e_target: float


class SignalBreakdownItem(BaseModel):
    name: str
    bullish: bool | None  # True=bullish, False=bearish, None=neutral/unavailable
    detail: str


class IchimokuAssessment(BaseModel):
    outlook: str  # "bullish" | "bearish" | "neutral"
    score: int
    max_score: int
    lookback_candles: int
    forecast_candles: int
    signals: list[SignalBreakdownItem]
    summary: str


class IchimokuResponse(BaseModel):
    symbol: str
    bars: list[HistoricalBar]
    points: list[IchimokuPoint]
    wave_targets: list[WaveTargetSet]
    assessment: IchimokuAssessment

File: frontend/src/types/ichimoku.ts
import type { HistoricalBar } from "@/types/market";

export interface IchimokuPoint {
  timestamp: string;
  tenkan: number | null;
  kijun: number | null;
  senkou_a: number | null;
  senkou_b: number | null;
  chikou: number | null;
  close: number | null;
}

export interface WavePivot {
  timestamp: string;
  price: number;
}

export interface WaveTargetSet {
  pivot_a: WavePivot;
  pivot_b: WavePivot;
  pivot_c: WavePivot;
  v_target: number;
  n_target: number;
  e_target: number;
}

export interface SignalBreakdownItem {
  name: string;
  bullish: boolean | null;
  detail: string;
}

export interface IchimokuAssessment {
  outlook: "bullish" | "bearish" | "neutral";
  score: number;
  max_score: number;
  lookback_candles: number;
  forecast_candles: number;
  signals: SignalBreakdownItem[];
  summary: string;
}

export interface IchimokuResponse {
  symbol: string;
  bars: HistoricalBar[];
  points: IchimokuPoint[];
  wave_targets: WaveTargetSet[];
  assessment: IchimokuAssessment;
}
